src/Foundation/NSUrlSessionHandler.cs

```diff
@@ -0,0 +1,716 @@
+				SetResponse(inflight);
+			}
+
+			public override void DidCompleteWithError(NSUrlSession session, NSUrlSessionTask task, NSError error)
+			{
+				var inflight = GetInflightData(task);
+
+				// this can happen if the HTTP request times out and it is removed as part of the cancelation process
+				if (inflight != null)
+				{
+					// set the stream as finished
+					inflight.Stream.TrySetReceivedAllData();
+
+					// send the error or send the response back 
+					if (error != null)
+					{
+						inflight.Errored = true;
+
+						var exc = CreateExceptionForNSError(error);
+						inflight.CompletionSource.TrySetException(exc);
+						inflight.Stream.TrySetException(exc);
+					}
+					else
+					{
+						inflight.Completed = true;
+						SetResponse(inflight);
+					}
+
+					RemoveInflightData(task);
+				}
+			}
+
+			private void SetResponse(InflightData inflight)
+			{
+				lock (inflight.Lock)
+				{
+					if (inflight.ResponseSent)
+						return;
+
+					if (inflight.CompletionSource.Task.IsCompleted)
+						return;
+
+					var httpResponse = inflight.Response;
+
+					inflight.ResponseSent = true;
+
+					// EVIL HACK: having TrySetResult inline was blocking the request from completing
+					Task.Run(() => inflight.CompletionSource.TrySetResult(httpResponse));
+				}
+			}
+
+			public override void DidReceiveChallenge(NSUrlSession session, NSUrlSessionTask task, NSUrlAuthenticationChallenge challenge, Action<NSUrlSessionAuthChallengeDisposition, NSUrlCredential> completionHandler)
+			{
+				if (challenge.ProtectionSpace.AuthenticationMethod == NSUrlProtectionSpace.AuthenticationMethodNTLM)
+				{
+					NetworkCredential credentialsToUse;
+
+					if (_handler.Credentials != null)
+					{
+						if (_handler.Credentials is NetworkCredential)
+						{
+							credentialsToUse = (NetworkCredential)_handler.Credentials;
+						}
+						else
+						{
+							var inflight = GetInflightData(task);
+							var uri = inflight.Request.RequestUri;
+
+							credentialsToUse = _handler.Credentials.GetCredential(uri, "NTLM");
+						}
+
+						var credential = new NSUrlCredential(credentialsToUse.UserName, credentialsToUse.Password, NSUrlCredentialPersistence.ForSession);
+						completionHandler(NSUrlSessionAuthChallengeDisposition.UseCredential, credential);
+					}
+
+					return;
+				}
+			}
+
+			public override void WillCacheResponse(NSUrlSession session, NSUrlSessionDataTask dataTask, NSCachedUrlResponse proposedResponse, Action<NSCachedUrlResponse> completionHandler)
+			{
+				// never cache
+				completionHandler(null);
+			}
+
+			public override void WillPerformHttpRedirection(NSUrlSession session, NSUrlSessionTask task, NSHttpUrlResponse response, NSUrlRequest newRequest, Action<NSUrlRequest> completionHandler)
+			{
+				var nextRequest = (_handler.AllowAutoRedirect ? newRequest : null);
+				completionHandler(nextRequest);
+			}
+
+			private static Exception CreateExceptionForNSError(NSError error)

```

Comment:
We used our own `createExceptionForNSError` method (in `HttpClientEx.cs`). The reasons were:

* to use our own `enums` (ModernHttpClient has its own because some values were missing in earlier versions);

* to hide CFNetwork usage on the watchOS profile - because it's not available there and we want to have a native handler on that platform;